hiredis.c

```diff
@@ -687,6 +688,14 @@ static redisContext *redisContextInit(const redisOptions *options) {
         return NULL;
 
     c->funcs = &redisContextDefaultFuncs;
+
+    /* Set any user supplied RESP3 PUSH handler or use freeReplyObject
```

Comment:
Doesn't this result with double free with the async context?

Reply:
No in [redisAsyncConnectWithOptions](https://github.com/redis/hiredis/pull/841/files#diff-8bd6782b369f6afec2c090b37d553d46R172) I flag us as not wanting a handler unless the user supplies one specifically.

There's still a danger of the user explicitly setting the handler to `freeReplyObject` but I'm not sure how we can avoid that.

Reply:
But `redisProcessCallbacks` calls `__redisRunPushCallback` (which calls `c->push_cb`, which ends up calling `freeReplyObject`), and then calls `c->reader->fn->freeObject` once again.

If the contract with the push callback is consistent with other callbacks, the callback gets a borrowed reply and must not free it. But then the default callback should be `NULL` and do nothing rather than `freeReplyObject`. Am I missing something here?

Reply:
There is every possibility I'm missing something :smile:

If you compile and run this demonstration program you'll find `push_cb` is NULL.

```c
#include "hiredis/hiredis.h"
#include "hiredis/async.h"

int main(void) {
    redisAsyncContext *ac;
    ac = redisAsyncConnect("localhost", 6379);

    /* ac->c.push_cb will be NULL */
    printf("Push callback: %p\n", ac->c.push_cb);
}
```

```bash
$ cc -o push-example async-push.c hiredis/libhiredis.a && ./push-example
Push callback: (nil)
```

It only defaults to `freeReplyObject` in synchronous hiredis but only so we don't mess up the order of replies that the caller sees with regular `redisCommand` usage.

If you think it better for the `push_cb` to be something like `void (pushHandler*)(const void *)` that's totally fine with me.  

The reason I don't automatically free the reply in the sync API is that it lets the user make the choice.  If they have a use case where they want to keep it around they can do so without having to copy memory.

Reply:
@michael-grunder you are right of course, my bad!